File: BinaryTree/11.BinaryTreeColoring(LeetCode).java
// Link :- https://leetcode.com/problems/binary-tree-coloring-game/submissions/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int xKaLeft;
    int xKaRight;
    
    public int size(TreeNode root , int x) {
        if(root == null) {
            return 0;
        }
        
        int ls = size(root.left , x);
        int rs = size(root.right , x);
        
        if(root.val == x) {
            xKaLeft = ls;
            xKaRight = rs;
        }
        
        int ts = ls + rs + 1;
        return ts;
    }
    public boolean btreeGameWinningMove(TreeNode root, int n, int x) {
        size(root , x);
        int theOtherSide = n - (xKaLeft + xKaRight + 1);
        int maxOfThree = Math.max(theOtherSide , Math.max(xKaLeft , xKaRight));
        int rest = n - maxOfThree;
        if(maxOfThree > rest){
            return true;
        }else{
            return false;
        }
    }
}

File: BinaryTree/11.BinaryTreeColoring(LeetCode).cpp
// Link :- https://leetcode.com/problems/binary-tree-coloring-game/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int xKaLeftSize;
    int xKaRightSize;
    
    int size(TreeNode * root , int x){
        if(root == NULL) {
            return 0;
        }
        
        int ls = size(root->left , x);
        int rs = size(root->right , x);
        
        if(root->val == x){
            xKaLeftSize = ls;
            xKaRightSize = rs;
        }
        
        int ts = ls + rs + 1;
        return ts; 
    }
    bool btreeGameWinningMove(TreeNode* root, int n, int x) {
        size(root , x);
        int sizeOfTheOtherSide = n - (xKaLeftSize + xKaRightSize + 1);
        int maxOfThree = max(sizeOfTheOtherSide , max(xKaLeftSize , xKaRightSize));
        int rest = n - maxOfThree;
        if(maxOfThree > rest) {
            return true;
        }else{
            return false;
        }
    }
};